server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java

```diff
@@ -0,0 +1,286 @@
+/**
+ * Copyright (c) 2017 Bosch Software Innovations GmbH.
+ * <p>
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ * <p>
+ * Contributors:
+ * Bosch Software Innovations GmbH - initial creation
+ */
+
+package org.eclipse.hono.service.credentials.impl;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.file.FileSystem;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import org.eclipse.hono.service.credentials.BaseCredentialsService;
+import org.eclipse.hono.util.CredentialsResult;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Repository;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static java.net.HttpURLConnection.*;
+import static org.eclipse.hono.util.CredentialsConstants.*;
+
+
+/**
+ * A credentials service that keeps all data in memory but is backed by a file.
+ * <p>
+ * On startup this adapter loads all added credentials from a file. On shutdown all
+ * credentials kept in memory are written to the file.
+ */
+@Repository
+@ConfigurationProperties(prefix = "hono.credentials")
+@Profile({"default", "credentials-file"})
+public class FileBasedCredentialsService extends BaseCredentialsService {
```

Comment:
Decided to mark the single operations for more flexibility. The currently still "not implemented" credentials API methods will remain not final but will be changed to final when implementation was added.